nbgrader/converters/base.py

```diff
@@ -248,6 +248,28 @@ def set_permissions(self, assignment_id, student_id):
         for dirname, _, filenames in os.walk(dest):
             for filename in filenames:
                 os.chmod(os.path.join(dirname, filename), permissions)
+            # If groupshared, set dir permissions - see comment below.
+            if self.coursedir.groupshared and os.stat(dirname).st_uid == os.getuid():
+                #for subdirname in subdirnames:
+                #    subdirname = os.path.join(dirname, subdirname)
+                try:
+                    os.chmod(dirname, (os.stat(dirname).st_mode|0o2770) & 0o2777)
+                except PermissionError:
+                    pass
+        # If groupshared, set write permissions on directories.  Directories
+        # are created within ipython_genutils.path.ensure_dir_exists via
+        # nbconvert.writer, (unless there are supplementary files) with a
+        # default mode of 755 and there is no way to pass the mode arguments
+        # all the way to there!  So we have to walk and fix.
+        if self.coursedir.groupshared:
+            # Root may be created in this step, and is not included above.
+            rootdir = self.coursedir.format_path(self._output_directory, '.', '.')
+            # Add 2770 to existing dir permissions (don't unconditionally override)
+            try:
+                os.chmod(rootdir, (os.stat(rootdir).st_mode|0o2770) & 0o2777)
+            except PermissionError:
+                pass
```

Comment:
If the permissions can't be set, wouldn't it be worthwhile to throw an error?

Reply:
For this (and all other permission errors / `ensure_root` comments), what I was thinking was that only the owner can change permissions - so any `chmod` call will always fail even if it is OK.  Since we are expecting this to be run by multiple users, failures will be the rule rather than the exception.

My thought for a proper solution now: add a test if permissions are OK and only try to `chmod` if needed.  If chmod is not possible, log a warning.

Another option is to add in some code comments / docs in the option explaining this and ignoring the problem for now.

Reply:
Oops, sorry I forgot to reply to this earlier. I see, that reasoning makes sense.

> My thought for a proper solution now: add a test if permissions are OK and only try to chmod if needed. If chmod is not possible, log a warning.

I think I would prefer that solution, I think silently hiding errors is always a bit dangerous 🙂

Reply:
I wasn't sure what to do, because it would fail in the common case when owner != current user.  In the end, I stat the file first, see if the mode needs to be changed, then try to change it - so errors will only come when it can't be changed _and_ it is actually a problem.  We had to stat everything anyway, so no major loss.

So I think these should be fixed now.